File: backend/src/kb_backend/models/answer.py
from datetime import date, datetime

from sqlalchemy import Boolean, CHAR, DateTime, ForeignKeyConstraint, Index, String, false
from sqlalchemy.dialects.mysql import BIGINT, JSON, LONGTEXT
from sqlalchemy.orm import Mapped, mapped_column

from .base import Base, created_at_column
from .knowledge_base import TABLE_ARGS


class Answer(Base):
    __tablename__ = "answer"
    __table_args__ = (
        ForeignKeyConstraint(
            ["knowledge_point_id", "knowledge_base_id"],
            ["knowledge_point.id", "knowledge_point.knowledge_base_id"],
            ondelete="RESTRICT",
            name="fk_answer_kp_kb",
        ),
        Index("ix_answer_resolve", "knowledge_point_id", "coord_hash", "effective_time", "created_at"),
        # Supports list_knowledge_points' per-KB active-answer-count query
        # (WHERE knowledge_base_id = ? AND revoked = 0 GROUP BY
        # knowledge_point_id) — neither index above leads with
        # knowledge_base_id. Added in migration 0003; found by the Kimi
        # review gate on PR #20.
        Index("ix_answer_kb_revoked_kp", "knowledge_base_id", "revoked", "knowledge_point_id"),
        TABLE_ARGS,
    )

    id: Mapped[int] = mapped_column(BIGINT(unsigned=True), primary_key=True, autoincrement=True)
    knowledge_base_id: Mapped[int] = mapped_column(BIGINT(unsigned=True), nullable=False)
    knowledge_point_id: Mapped[int] = mapped_column(BIGINT(unsigned=True), nullable=False)
    coord: Mapped[dict] = mapped_column(JSON, nullable=False)
    coord_hash: Mapped[str] = mapped_column(CHAR(64), nullable=False)
    content: Mapped[str] = mapped_column(LONGTEXT, nullable=False)
    effective_time: Mapped[date] = mapped_column(nullable=False)
    operator: Mapped[str] = mapped_column(String(100), nullable=False, server_default="admin")
    # 数据来源（产生方式）：人工填报 / AI生成 / 批量导入（migration 0008 起
    # 收敛为三值；历史值"人工编辑"已归一为"人工填报"）
    source: Mapped[str] = mapped_column(String(100), nullable=False, server_default="人工填报")
    # 操作系统（写入方系统编码）：服务凭证写入 = 平台来源系统编码（bqxt/
    # yhfkglxt…），运营端/本系统写入 = tyzsk。取值见 auth.deps.current_source_system
    source_system: Mapped[str] = mapped_column(String(100), nullable=False, server_default="tyzsk")
    note: Mapped[str | None] = mapped_column(LONGTEXT, nullable=True)
    revoked: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default=false())
    revoked_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    revoked_by: Mapped[str | None] = mapped_column(String(100), nullable=True)
    revoke_reason: Mapped[str | None] = mapped_column(String(500), nullable=True)
    # 重新启用三件套（issue #32）：与 revoke 三件套同构的链级字段。恢复时
    # revoked_* 保留原样当历史，只保留最近一轮撤回/恢复（完整事件史留 V2，
    # 见 docs/specs/2026-08-12-answer-reactivate-design.md §1）。
    reactivated_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    reactivated_by: Mapped[str | None] = mapped_column(String(100), nullable=True)
    reactivate_reason: Mapped[str | None] = mapped_column(String(500), nullable=True)
    created_at = created_at_column()

File: frontend/src/api/answer.ts
/**
 * 答案（字段镜像 backend/src/kb_backend/schemas/knowledge_point.py::AnswerOut）。
 * 参数清洗（trim、空转 undefined）由调用方页面在提交前完成后传入。
 */
import type { FilterValue } from '@/utils/dimension'
import { request } from '@/utils/request'

export interface Answer {
  id: number
  knowledge_base_id: number
  knowledge_point_id: number
  coord: Record<string, string | number | boolean>
  coord_hash: string
  content: string
  effective_time: string
  operator: string
  /** 数据来源（产生方式）：人工填报 / AI生成 / 批量导入 */
  source: string
  /** 操作系统（写入方系统编码）：tyzsk = 本系统，其他为外部系统编码 */
  source_system: string
  note: string | null
  revoked: boolean
  revoked_at: string | null
  revoked_by: string | null
  revoke_reason: string | null
  reactivated_at: string | null
  reactivated_by: string | null
  reactivate_reason: string | null
  created_at: string
}

export interface CreateAnswerInput {
  coord: Record<string, FilterValue>
  content: string
  effective_time: string
  note?: string
  /** issue #32：目标条件组合当前是撤回态时必填——写入即整链恢复 + 追加新版本 */
  reactivate_reason?: string
}

export interface EditAnswerInput {
  content: string
  effective_time: string
  note?: string
  /**
   * 设计文档 §4.4：条件没动过时整个省略（连原值都不发），让后端走
   * 「原样复用目标 coord、跳过重校验」的路径——这条路径是「答案自身 coord
   * 引用了已停用维度时仍能编辑内容/时间」得以成立的原因。
   */
  coord?: Record<string, FilterValue>
  migration_reason?: string
  /** issue #32：编辑落点的链（迁移后的新条件或原条件本身）是撤回态时必填 */
  reactivate_reason?: string
}

/** 写一条答案（4.x） */
export function createAnswer(kbId: number, kpId: number, input: CreateAnswerInput) {
  return request.post<Answer>(`/knowledge-bases/${kbId}/knowledge-points/${kpId}/answers`, input)
}

/** 编辑答案（同组追加新版本；coord 变更即迁移） */
export function editAnswer(kbId: number, kpId: number, answerId: number, input: EditAnswerInput) {
  return request.post<Answer>(`/knowledge-bases/${kbId}/knowledge-points/${kpId}/answers/${answerId}/edit`, input)
}

/** 撤回答案（逻辑删除，历史与留痕永久保留） */
export function revokeAnswer(kbId: number, kpId: number, answerId: number, revokeReason: string) {
  return request.post<Answer>(`/knowledge-bases/${kbId}/knowledge-points/${kpId}/answers/${answerId}/revoke`, {
    revoke_reason: revokeReason,
  })
}
